OvmfPkg/PlatformPei/PlatformId.c

```diff
@@ -0,0 +1,660 @@
+      // not NUL-terminated. Add a NUL terminator to print as a string.
+      ByteStrToCstr (Buffer, sizeof (Buffer), Locator, LocatorEnd);
+      DEBUG ((DebugLevel, DPREFIX "URI locator: %a\n", Buffer));
+      break;
+    case LOCATOR_TYPE_DEVICE_PATH: // EFI Device Path protocol binary representation.
+      BytesToHex (Buffer, sizeof (Buffer), Locator, LocatorEnd);
+      DEBUG ((DebugLevel, DPREFIX "Device path locator (to-text protocol not available): %s\n", Buffer));
+      break;
+    case LOCATOR_TYPE_UEFI_VARIABLE:
+      DEBUG ((DebugLevel, DPREFIX "UEFI variable locator: %s-%g\n", Locator + sizeof (EFI_GUID), Locator));
+      break;
+    default:
+      DEBUG ((DebugLevel, DPREFIX "Unknown locator type: %d\n", LocatorType));
+      break;
+  }
+}
+
+STATIC
+VOID
+DebugCursors (
+  IN UINTN                                    DebugLevel,
+  IN TCG_Sp800_155_PlatformId_Event3_Cursors  *Cursors
+  )
+{
+  DEBUG ((DebugLevel, DPREFIX "VendorId: %d\n", Cursors->VendorId));
+  DEBUG ((DebugLevel, DPREFIX "ReferenceManifestGuid: %g\n", Cursors->ReferenceManifestGuid));
+  DEBUG ((DebugLevel, DPREFIX "PlatformManufacturerStr: %a\n", Cursors->PlatformManufacturerStr));
+  DEBUG ((DebugLevel, DPREFIX "PlatformModel: %a\n", Cursors->PlatformModel));
+  DEBUG ((DebugLevel, DPREFIX "PlatformVersion: %a\n", Cursors->PlatformVersion));
+  DEBUG ((DebugLevel, DPREFIX "FirmwareManufacturerStr: %a\n", Cursors->FirmwareManufacturerStr));
+  DEBUG ((DebugLevel, DPREFIX "FirmwareManufacturerId: %d\n", Cursors->FirmwareManufacturerId));
+  DEBUG ((DebugLevel, DPREFIX "FirmwareVersion: %a\n", Cursors->FirmwareVersion));
+  DebugLocator (DebugLevel, Cursors->RimLocator, Cursors->RimLocatorEnd, Cursors->RimLocatorType);
+  DebugLocator (DebugLevel, Cursors->PlatformCertLocator, Cursors->PlatformCertLocatorEnd, Cursors->PlatformCertLocatorType);
+}
+
+/**
+ * Creates an EFI_HOB_TYPE_GUID_EXTENSION HOB for each SP800155  event in the
+ * span [Evts, Evts + EvtsSize). Associates the string data with
+ * gTcg800155PlatformIdEventHobGuid. Any unused bytes or out-of-bounds event
+ * sizes are considered corrupted and are discarded.
+**/
+STATIC
+VOID
+PlatformIdRegisterSp800155 (
+  IN CONST EFI_PEI_SERVICES  **PeiServices,
+  IN       UINT8             *Evts,
+  IN       UINTN             EvtsSize
+  )
+{
+  EFI_STATUS                               Status;
+  UINTN                                    EvtSize;
+  VOID                                     *Hob;
+  EFI_HOB_GUID_TYPE                        *GuidHob;
+  UINT8                                    *Evt;
+  UINT8                                    *EvtNext;
+  UINT8                                    *EvtsEnd;
+  UINT8                                    *EvtDest;
+  TCG_Sp800_155_PlatformId_Event3_Cursors  Data;
+
+  Evt     = Evts;
+  EvtsEnd = Evts + EvtsSize;
+  while (Evt != EvtsEnd) {
+    Status = ParseSp800155 (Evt, EvtsEnd, &Data);

```

Comment:
I thought so too, but the variable length without a length header requires all the cursor logic. I had a simpler implementation but @kraxel wanted this instead.

Reply:
To me, I always like simple and obvious code. Complicated code == Risky and less secure.

Reply:
Yes.  I didn't like the idea to have an additional header wrapped around the struct, that makes the handling more complicated.  Going for one file per record (instead of concating them) is indeed a more elegant solution to the problem.

Another concern is that OVMF might want apply sanity checks to the record, such as checking that the version in the record actually matches `PcdFirmverVersionString` (not implemented in this PR but easily doable as incremental change with the cursors).

Reply:
Removed for uninterpreted single file solution.